fix(core): use custom data attribute as identifyer for test components

json-derulo opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/main/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.dev application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Currently, test components are identified by adding an `id` attribute to the component's DOM. This method fails to identify components if the host's ID is overridden via host binding, resulting in components not being cleaned up.

Issue Number: Closes #35215 


## What is the new behavior?

 The new approach uses a custom `data-ng-test-id` attribute, preventing such overrides and ensuring proper identification and cleanup of test components.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

JeanMeche commented on packages/platform-browser-dynamic/testing/src/dom_test_component_renderer.ts:
```diff
@@ -22,7 +22,7 @@ export class DOMTestComponentRenderer extends TestComponentRenderer {
     this.removeAllRootElementsImpl();
     const rootElement = getDOM().getDefaultDocument().createElement('div');
-    rootElement.setAttribute('id', rootElId);
+    rootElement.setAttribute('data-ng-test-id', rootElId);
```
Nit: maybe add a comment to why we use such a "complex" attribute name.

json-derulo committed:
fix(core): use custom data attribute as identifyer for test components

Currently, test components are identified by adding an `id` attribute to the component's DOM. This method fails to identify components if the host's ID is overridden via host binding, resulting in components not being cleaned up. The new approach uses a custom `data-ng-test-id` attribute, preventing such overrides and ensuring proper identification and cleanup of test components.

crisbeto commented on packages/platform-browser-dynamic/testing/src/dom_test_component_renderer.ts:
```diff
@@ -22,7 +22,9 @@ export class DOMTestComponentRenderer extends TestComponentRenderer {
-    rootElement.setAttribute('id', rootElId);
+    // This verbose attribute name is chosen to minimize the chances
+    // of collision with a host binding of the tested component
+    rootElement.setAttribute('data-ng-test-id', rootElId);
```
It's worth noting that `data-` attributes seem to cause some slowdowns as per #59264.

json-derulo replied:
Thanks for pointing that out, however I think that in this case the effect of the slowdown is very small. The element has just been created in the line before, and the element isn't added to the DOM yet, so setting the attribute shouldn't be an expensive operation. In ADEV it seems like it has triggered some change detection and re-painting, which is not relevant here. I just tried some benchmarks in my browser - I couldn't see any significant slowdown when using `data-*` attributes.

JeanMeche replied:
The issue on ADEV was that we also created a large number of those nodes at once.

json-derulo commented:
The ADEV tests are failing. The problem is that for `@angular/platform-browser-dynamic` the updated code is used, but for `@angular/core/testing` still the old code is used.

Does anyone have an idea how to fix it? Maybe @JeanMeche?

JeanMeche commented:
We've had similar issues in the past, unfortunately there is no real way to work around this beside disabling the test. 

What we've done previously is disable the test and merge the day of the release and re-enable the test once the release is out and we can update Adev's dependencies.

JeanMeche commented:
The ADEV tests are currently disabled, let's try to land this !

thePunderWoman reviewed (approved):
LGTM

mmalerba commented:
This has failures in google3 that need to be investigated & addressed before it can land

JeanMeche commented:
I just had a look at the breakages. 

It boils down to several project not using the `DOMTestComponentRenderer` but their own custom `TestComponentRenderer`. 

In result `componentFactory.create` throws 

https://github.com/json-derulo/angular/blob/c44b1cd84c0ec0e84d3c323f93063dae10582c60/packages/core/testing/src/test_bed.ts#L652-L660

The throwing location: 
https://github.com/angular/angular/blob/75520eb7f7da5efcfaf8e0dbd09e8f03a98dbaa1/packages/platform-browser/src/dom/dom_renderer.ts#L323-L332
